addons/membership/models/membership.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+import odoo.addons.decimal_precision as dp
+
+STATE = [
+    ('none', 'Non Member'),
+    ('canceled', 'Cancelled Member'),
+    ('old', 'Old Member'),
+    ('waiting', 'Waiting Member'),
+    ('invoiced', 'Invoiced Member'),
+    ('free', 'Free Member'),
+    ('paid', 'Paid Member'),
+]
+
+
+class MembershipLine(models.Model):
+    _name = 'membership.membership_line'
+    _description = __doc__
+    _rec_name = 'partner'
+    _order = 'id desc'
+
+    partner = fields.Many2one('res.partner', string='Partner', ondelete='cascade', index=True)
+    membership_id = fields.Many2one('product.product', string="Membership", required=True)
+    date_from = fields.Date(string='From', readonly=True)
+    date_to = fields.Date(string='To', readonly=True)
+    date_cancel = fields.Date(string='Cancel date')
+    date = fields.Date(string='Join Date',
+        help="Date on which member has joined the membership")
+    member_price = fields.Float(string='Membership Fee',
+        digits_compute=dp.get_precision('Product Price'), required=True,
+        help='Amount for the membership')
+    account_invoice_line = fields.Many2one('account.invoice.line', string='Account Invoice line', readonly=True, ondelete='cascade')
+    account_invoice_id = fields.Many2one('account.invoice', related='account_invoice_line.invoice_id', string='Invoice', readonly=True)
+    company_id = fields.Many2one('res.company', related='account_invoice_line.invoice_id.company_id', string="Company", readonly=True, store=True)
+    state = fields.Selection(STATE, compute='_compute_state', string='Membership Status', store=True,
+        help="It indicates the membership status.\n"
+             "-Non Member: A member who has not applied for any membership.\n"
+             "-Cancelled Member: A member who has cancelled his membership.\n"
+             "-Old Member: A member whose membership date has expired.\n"
+             "-Waiting Member: A member who has applied for the membership and whose invoice is going to be created.\n"
+             "-Invoiced Member: A member whose invoice has been created.\n"
+             "-Paid Member: A member who has paid the membership amount.")
+
+    @api.depends('account_invoice_line.invoice_id.state','partner.membership_state')
```

Comment:
I don't see where `partner.membership_state` is used, however line 76 doesn't seem reflected in the dependencies.